### DISCUSSION TITLE
내부 반복과 외부 반복 그리고 중간 연산과 최종 연산

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#48] [CHAPTER 4 스트림 소개 - Q&A]
> 이번 챕터는 스트림API 와 컬렉션 API 의 동작 차이에 대해 전반적으로 설명해준다. 
> 자세한 내용은 뒤에 챕터에 설명되므로 전반적으로 이해한 내용이 맞는지 정리하려고 한다.

- 내부 반복과 외부 반복에 대해서 책에서 설명과 중간 연산과 최종 연산에 대한 설명이 있다.
- 중간 연산은 스트림으로 반환하고 지연(lazily) 연산을 하게된다. 따라서 최종 연산이 있기 전까지 실제 연산되지 않고, 최종 연산을 할 때 스트림API 는 비로소 한 흐름으로 연산을 실행하고 끝마치게 된다.

Q. 스트림 API 의 최종 연산인 `forEach` 와 `count`, `collect` 등 은 외부 동작이고,
중간 연산인 `filter`, `map`, `limit`, `sorted`, `distinct` 등 중간 연산은 내부 동작이라고 볼 수 있는 것일까??
(아래 그림처럼)

<img width="384" alt="image" src="https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/71637466/b22d9d6e-daa9-422c-a293-7a456354e425">

   

--- 

그렇다면,  최종 연산(외부 연산)을 하게 되면서 스트림(커다란 하나의 흐름?)이 끝나기 때문에 [스트림 API 의 일회성](https://github.com/orgs/Study-2-Modern-Java-In-Action/discussions/46) 이 나타나는 건지?

**Comment by JoisFe** [comment]
그렇지 않나요?
[스트림 API 의 일회성](https://github.com/orgs/Study-2-Modern-Java-In-Action/discussions/46)
해당 링크가 스트림 API의 일회성을 잘 설명해주는 것 같습니다.
